### PR TITLE
feat: add importer component shutdown and init by method

### PR DESCRIPTION
**Pull request by gabriel-aranha-cw**
### **PR Type**
Enhancement, Tests

___

### **Description**
- Added importer shutdown and initialization functionality:
  - New methods `stratus_init_importer` and `stratus_shutdown_importer` in RPC server
  - Global importer shutdown control in `globals.rs`
  - Enhanced error handling for importer-related operations
- Improved consensus handling:
  - Updated `RpcContext` to use `RwLock` for thread-safe consensus access
  - Modified error types and handling related to consensus
- Enhanced end-to-end testing:
  - New comprehensive test suite for importer functionality
  - Updated GitHub Actions workflow to run tests for both BRLC and importer
- Added utility classes and functions for testing:
  - New `Account` class for transaction signing
  - Constants and utility functions for chain ID and hex conversion
- Updated build and test scripts for better flexibility and error handling

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>importer.rs</strong><dd><code>Enhance importer shutdown logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/follower/importer/importer.rs

<li>Added <code>should_shutdown</code> method to check if importer should shutdown<br> <li> Updated shutdown checks in various methods to use <code>should_shutdown</code><br> <li> Improved error handling and logging in several methods<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-bd50f502f039aff8f75753bc780d281ae7d7936fadc9d53482084c647aec94ca">+19/-10</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>importer_config.rs</strong><dd><code>Add deserialization support for ImporterConfig</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/follower/importer/importer_config.rs

- Added `serde::Deserialize` derive for `ImporterConfig` struct

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-d6208b041b3fff7e5d6e8ed530417cc9ac4d6037a7a9737370e05322418d1d34">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>miner_config.rs</strong><dd><code>Update genesis block creation logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/miner/miner_config.rs

- Modified condition for creating genesis block and accounts

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-ea3a8597b909e696993f7e791ab04e5ad784da8077b87088d9c78c804e1ef69d">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>stratus_error.rs</strong><dd><code>Expand error types for Importer and Consensus</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/primitives/stratus_error.rs

<li>Added new error types for Importer, Consensus, and AppConfig<br> <li> Removed <code>RpcConsensusUnavailable</code> error<br> <li> Added <code>StratusNotFollower</code> error<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-301cded01d772388e3e5c08ede093b91c3e4478c08e11a48f70e3d44351385fb">+45/-4</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rpc_context.rs</strong><dd><code>Update RpcContext to use RwLock for consensus</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/eth/rpc/rpc_context.rs

- Changed `consensus` field to use `RwLock<Option<Arc<dyn Consensus>>>`

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-a486ea1bd70eb81a959b89f514ed248a5c30fb4dabe7c08344b9bac246b6e92e">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rpc_server.rs</strong><dd><code>Add importer control methods and improve consensus handling</code></dd></summary>
<hr>

src/eth/rpc/rpc_server.rs

<li>Added new methods <code>stratus_init_importer</code> and <code>stratus_shutdown_importer</code><br> <li> Updated error handling in <code>stratus_health</code> and <code>eth_send_raw_transaction</code><br> <li> Modified consensus access to use RwLock<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-835ba255c9a5c1fe0e13285bc39e058e1c74422e5e03ebba483c9d8a15c45405">+122/-18</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>globals.rs</strong><dd><code>Add global importer shutdown control</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/globals.rs

<li>Added <code>IMPORTER_SHUTDOWN</code> static variable<br> <li> Implemented methods for importer shutdown control<br> <li> Updated <code>GlobalState</code> struct to include <code>is_importer_shutdown</code><br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-48d6fbc3a4ed67100952dcccfada858623c931e73d6de32984d4d1457e68d3a9">+44/-2</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rpc.ts</strong><dd><code>Add chain ID constants and hex conversion utility</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

e2e/cloudwalk-contracts/integration/test/helpers/rpc.ts

<li>Added <code>CHAIN_ID_DEC</code> and <code>CHAIN_ID</code> constants<br> <li> Implemented <code>toHex</code> function for number to hex conversion<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-bfbd0a045ec4243ba68015b78cb7fb7cb485ff47e44bc3c22e3e155fc3ddf003">+8/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>e2e-leader-follower.yml</strong><dd><code>Update GitHub Actions workflow for multiple tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/e2e-leader-follower.yml

<li>Added matrix strategy to run tests for both BRLC and importer<br> <li> Updated job name and concurrency group to include test type<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-6cb08ce033aeec24672fd91944789df5188b47bbf56f16d034c600ba66ea1461">+5/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>justfile</strong><dd><code>Update e2e test recipe for flexibility</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

justfile

<li>Modified <code>e2e-leader-follower-up</code> recipe to accept test parameter<br> <li> Updated test command to use specific test file based on parameter<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-deb9bb56fb122db0b605aa5b63f95a4665c905b18dd670e1fa6c877576a94ff1">+2/-12</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>account.ts</strong><dd><code>Add Account class and test account utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

e2e/cloudwalk-contracts/integration/test/helpers/account.ts

<li>Added new <code>Account</code> class with methods for signing transactions<br> <li> Defined test accounts and utility function for generating random <br>accounts<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-e49702c7915bb9c0fffc7eb3704c77c2b2eda8c6509d82d638b91dcf87b9a7e2">+93/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>leader-follower-brlc.test.ts</strong><dd><code>Update BRLC test description</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

e2e/cloudwalk-contracts/integration/test/leader-follower-brlc.test.ts

- Updated test description to specify BRLC integration test

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-9c6d7de0f652c0d9179b3a178bc7626eccf0bc0a5bc3402b2dd4d910f9fc359a">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>leader-follower-importer.test.ts</strong><dd><code>Add comprehensive importer integration tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

e2e/cloudwalk-contracts/integration/test/leader-follower-importer.test.ts

<li>Implemented comprehensive test suite for importer functionality<br> <li> Added tests for importer shutdown, initialization, and transaction <br>handling<br> <li> Included utility functions for waiting and transaction confirmation<br>

</details>

  </td>
  <td><a href="https://github.com/cloudwalk/stratus/pull/1650/files#diff-76e205db03980cadded8f3f9ff7be76d1f3d3af18d4d8bf2cc8fa09d74934b8d">+244/-0</a>&nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: src/eth/rpc/rpc_server.rs [hunk 1]
```diff
@@ -273,6 +282,96 @@ fn stratus_reset(_: Params<'_>, ctx: Arc<RpcContext>, _: Extensions) -> Result<J
...
+        return false;
+    }
+
+    if GlobalState::is_importer_shutdown() {
+        return true;
+    }
+
+    const TASK_NAME: &str = "rpc-server::importer-shutdown";
+    GlobalState::shutdown_importer_from(TASK_NAME, "received importer shutdown request");
+    GlobalState::is_importer_shutdown()
```
Comment (marcospb19-cw) [inline src/eth/rpc/rpc_server.rs:372]: nitpick: this could return true directly.

Comment (gabriel-aranha-cw) [reply src/eth/rpc/rpc_server.rs:372]: Thanks. I still have on my to do list some code cleanup and small refactor. I will add it to the list

Comment (gabriel-aranha-cw) [reply src/eth/rpc/rpc_server.rs]: Done.

**Pull request merged**